cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,79 +284,431 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+    v                       v                             v                        |              |
++---+---+           +-------+----+                +-------+----+                   |              |
+|       |           | mount      |       success  | mount data |                   |              |
+| done  +<----------+ save       |      +---------+            +<------------------+              |
+|       |           |            |      |         |            |                                  |
++--+----+           +----+-------+      |         +----------+-+                                  |
+   ^                     ^              |                    |                                    |
+   |                     |success       v                    |                                    |
+   |                     |     +--------+----+   fail        |fail                       unlock   |
+   |                     |     | unlock save +--------+      |                           failed   |
+   |                     +-----+ w/ run key  |        v      v                                    |
+   |                     ^     +-------------+   +----+------+-----+                              |
+   |                     |                       | unlock save     |                              |
+   |                     |                       | w/ fallback key +<-----------------------------+
+   |                     +-----------------------+                 |
+   |                             success         +-------+---------+
+   |                                                     |
+   |                                                     |
+   |                                                     |
+   +-----------------------------------------------------+
+                         fail
+
+*/
+type stateMachine struct {
+	current stateFunc
+}
+
+func newStateMachine() *stateMachine {
+	m := &stateMachine{}
+	m.current = m.unlockDataRunKey
+	return m
+}
+
+func (m *stateMachine) execute(ctx *recoverContext) (finished bool, err error) {
+	next, err := m.current(ctx)
+	m.current = next
+	return next == nil, err
+}
+
+// stateUnlockDataRunKey will try to unlock ubuntu-data with the normal run-mode
+// key, and if it fails, progresses to the next state, which is either:
+// - failed to unlock data, but we know it's an encrypted device -> try to unlock with fallback key
+// - failed to find data at all -> try to unlock save
+// - unlocked data with run key -> mount data
+func (m *stateMachine) unlockDataRunKey(ctx *recoverContext) (stateFunc, error) {
 	runModeKey := filepath.Join(boot.InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key")
-	device, isDecryptDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(disk, "ubuntu-data", runModeKey, lockKeysOnFinish)
+	dataDevice, isEncryptedDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(ctx.disk, "ubuntu-data", runModeKey, false)
```

Comment:
I  don't understand how this is supposed to work, done this way this will ask the recovery key before trying the fallback key, no?

Reply:
Yeah, looks like  `UnlockVolumeUsingSealedKeyIfEncrypted` does too much and falls back to recovery key automatically. We should export secboot.UnlockEncryptedPartitionWithRecoveryKey (maybe s/Partition/Volume/ to be consistent?)

Reply:
this is fixed now